### DISCUSSION TITLE
Using MIDI controller buttons with MiniDexed

### DISCUSSION THREAD

**Discussion by diyelectromusic** [probonopd/MiniDexed#360] [Ideas]
Is this the kind of thing that might be of interest to people?  For more, see discussion here: https://github.com/probonopd/MiniDexed/issues/316

I have an experimental branch of MiniDexed that starts to do it, but would like a bit of guidance as to how useful that might be before doing anything else on it :) If this is of interest and looks like it might work, then it would be nice to get this included before the branch gets too far out of date, but it would have to be considered "experimental" for now :)

If you're interested and can build MiniDexed yourself to try it and give me some feedback, see: https://github.com/diyelectromusic/MiniDexed/tree/buttonmidi

Kevin

**Comment by BobanSpasic** [comment]
Hi @diyelectromusic 
would it be too much work to get these settings into a separate file?
I would suggest something like MIDI_Buttons_01.cfg and just a reference/link in minidexed.ini to the default or to the last configuration.
The other possibility would be to have the default or the last configuration in minidexed.ini, and loadable configurations as separate files.

This way, in the future, we could have configurations that could be loaded from the menu (if one has more than one master keyboard or for an exchange of configurations between the users).

**Reply by probonopd** [reply]
My 2 cents: We should have just one config file for all settings in order to keep things simple for _users_. As soon as you have to edit multiple files, the chance for errors increases and things are harder to document/support.

**Reply by diyelectromusic** [reply]
I agree with @probonopd.
For now anyway, MiniDexed is quite tied to its IO configuration in one place, so whilst it could be a possibility in the future (mirroring what happens with the performance.ini for example) I guess, that is a more significant structural change.

In fact, it would probably make more sense to have the whole MIDI configuration in one separate place so that all CCs, Notes, program messages, and so on could be separately configurable or loadable. But as I say, I think that is a change for another day :)

**Comment by probonopd** [comment]
> If you're interested and can build MiniDexed yourself to try it and give me some feedback, see: https://github.com/diyelectromusic/MiniDexed/tree/buttonmidi

...and if you're interested and _can't_ build MiniDexed yourself you can use

[MiniDexed_2022-10-03-2093704.zip](https://github.com/probonopd/MiniDexed/suites/8588566588/artifacts/384698190) from https://github.com/probonopd/MiniDexed/pull/365.

**Comment by probonopd** [comment]
Thanks a lot @diyelectromusic, finally found some time to test it.

My test results: 

* Can it be that `MIDIButtonPrev` and `MIDIButtonNext` are mixed up?
* On my keyboard, the buttons toggle the value between 0 and 127. If I press the button, then either the value 0 or the value 127 gets sent. Currently I have to press twice before the button press gets recognized. Can it be changed so that single button presses are recognized? Currently i see `uibotton: Event: 2` only if double click in a certain speed, even though I have set all actions to `click`. Example of the MIDI messages:
```
MIDI0: B0 02 7F
MIDI0: B0 02 00
uibotton: Event: 2 # only if double click in a certain speed

# Pressing just once
MIDI0: B0 02 7F
# uibotton: Event: 2 never appears

# Double clicking in the same speed but starting from the other value
MIDI0: B0 02 00
MIDI0: B0 02 7F
# uibotton: Event: 2 never appears
```

Here is my configuration:

```
# GPIO Button Navigation
#  Any buttons set to 0 will be ignored
ButtonPinPrev=0
ButtonActionPrev=click
ButtonPinNext=0
ButtonActionNext=click
ButtonPinBack=0
ButtonActionBack=click
ButtonPinSelect=0
ButtonActionSelect=click
ButtonPinHome=0
ButtonActionHome=click
ButtonPinShortcut=0
# (Shortcut doesn't have an action)

# Timeouts in milliseconds for double click and long press
DoubleClickTimeout=400
LongPressTimeout=400

# MIDI Button Navigation
#  Specify MIDI CC to act as a button
#  NB: Off < 64 < ON
# CC channel: 0=OFF; 1-16 MIDI Ch; >16 Omni
MIDIButtonCh=17
MIDIButtonPrev=00
MIDIButtonNext=02
MIDIButtonHome=06
MIDIButtonSelect=04
MIDIButtonBack=03
```

**Reply by diyelectromusic** [reply]
I expect we could decide to do whatever we like.  I suspect there are many, many configuration choices when setting up how MIDI buttons work, so we probably need to decide what should be config on the MiniDexed side and what should be config on the MIDI controller side.  At the moment though, it is mirroring how an actual hardware button would work which is simplest for the code.

Another idea would be to allow NoteOn/Off messages to act as buttons - that is another way some controllers work.

I'll look at Prev/Next! :)

Kevin

**Reply by probonopd** [reply]
> what should be config on the MiniDexed side and what should be config on the MIDI controller side

Thanks, that was the hint I needed! Indeed it is possible to change the behavior of the buttons on my MIDI controller. When I set them to "MIDI CC Trigger/Release", then it works as is.

![image](https://user-images.githubusercontent.com/2480569/193652890-1529f933-c1fe-4d7b-8e66-c18e88542a8f.png)

**Reply by diyelectromusic** [reply]
Ah yes, having looked through the code again, I can see that in userinterface.cpp the call to create a new CUIButton object has MIDINext/MIDIPrev in the wrong order as parameters :)

I've updated it :)

Thanks!
Kevin

**Reply by probonopd** [reply]
Works :+1: Ready to merge?

**Reply by diyelectromusic** [reply]
Up to you I guess... As a config option, I guess it can't hurt and if it starts disabled, then it shouldn't break too much :)

Maybe it is time to start thinking about a develop branch and a stable branch.  MiniDexed is a bit of a moving platform at the moment!

But as long as people are clear it is constantly changing, then probably not an issue!

Kevin

**Reply by probonopd** [reply]
Merged. Thank you very much! :+1:

**Comment by diyelectromusic** [comment]
OK, so here is the first go with an update that hopefully will allow you to use NoteOn/NoteOff messages as "MIDI buttons" too.  Just turn it on with:
MIDIButtonNotes=1
and then set the MIDIButton values to MIDI note numbers instead - e.g. 60, 62, 64, 65, 67 for C4-G4.

Of course if you're using a MIDI channel that you're playing on then things will get, er, interesting...

See: https://github.com/diyelectromusic/MiniDexed/tree/buttonmidi2

Kevin

**Reply by probonopd** [reply]
That's a clever idea - so now we can use the white keys on MIDI controllers that don't have extra buttons? :+1:

**Reply by diyelectromusic** [reply]
Yes, can use any keys/buttons/notes or whatever now - as I say, just don't get them mixed up with notes you'd actually want to play :)

**Reply by probonopd** [reply]
Opened a PR: https://github.com/probonopd/MiniDexed/pull/368
